File: Graph_Functions/Dijkstra_s/Dijkstra.java
import java.util.*;

public class Dijkstra {

    static class Pair implements Comparable<Pair> {
        int node;
        int weight;

        Pair(int node, int weight) {
            this.node = node;
            this.weight = weight;
        }

        @Override
        public int compareTo(Pair o) {
            return Integer.compare(this.weight, o.weight);
        }
    }

    static List<Integer> dijkstra(int V, List<List<Pair>> graph, int S) {

        PriorityQueue<Pair> pq = new PriorityQueue<>();
        List<Integer> dist = new ArrayList<>(Collections.nCopies(V, Integer.MAX_VALUE));

        pq.add(new Pair(S, 0));
        dist.set(S, 0);

        while (!pq.isEmpty()) {
            Pair current = pq.poll();
            int node = current.node;
            int dis = current.weight;

            for (Pair neighbor : graph.get(node)) {
                int adjNode = neighbor.node;
                int edgeWeight = neighbor.weight;

                if (dis + edgeWeight < dist.get(adjNode)) {
                    dist.set(adjNode, dis + edgeWeight);
                    pq.add(new Pair(adjNode, dist.get(adjNode)));
                }
            }
        }
        return dist;
    }

    public static void main(String[] args) {
        int V = 4, S = 2;
        List<List<Pair>> graph = new ArrayList<>();

        for (int i = 0; i < V; i++) {
            graph.add(new ArrayList<>());
        }

        graph.get(0).add(new Pair(1, 1));
        graph.get(0).add(new Pair(2, 6));
        graph.get(1).add(new Pair(2, 3));
        graph.get(1).add(new Pair(0, 1));
        graph.get(2).add(new Pair(1, 3));
        graph.get(2).add(new Pair(0, 6));

        List<Integer> res = dijkstra(V, graph, S);

        for (int i = 0; i < V; i++) {
            System.out.print("Shortest distance from " + S + " to " + i + " is: ");
            if (res.get(i) == Integer.MAX_VALUE) {
                System.out.println("INF");
            }
            else {
                System.out.println(res.get(i));
            }
        }
    }
}

File: Graph_Functions/Dijkstra_s/Dijkstra.py
import heapq

def dijkstra(V, graph, S):

    for neighbors in graph:
        for edge in neighbors:
            if edge[1] < 0:
                return [-1] * V

    dist = [float('inf')] * V
    dist[S] = 0
    pq = [(0, S)]

    while pq:
        dis, node = heapq.heappop(pq)

        if dis > dist[node]:
            continue

        for adj_node, edge_weight in graph[node]:
            if dis + edge_weight < dist[adj_node]:
                dist[adj_node] = dis + edge_weight
                heapq.heappush(pq, (dist[adj_node], adj_node))

    return dist

if __name__ == "__main__":
    V = 4
    S = 2
    graph = [[] for _ in range(V)]

    graph[0].append((1, 1))
    graph[0].append((2, 6))
    graph[1].append((2, 3))
    graph[1].append((0, 1))
    graph[2].append((1, 3))
    graph[2].append((0, 6))

    res = dijkstra(V, graph, S)

    for i in range(V):
        print(f"Shortest distance from {S} to {i} is: ", end="")
        if res[i] == float('inf'):
            print("INF")
        else:
            print(res[i])